pkg/auth/dependency/interaction/provider_action.go

```diff
@@ -110,6 +112,48 @@ func (p *Provider) performActionUpdateIdentity(i *Interaction, intent *IntentUpd
 	panic("interaction_add_identity: unhandled step " + step.Step)
 }
 
+func (p *Provider) performActionUpdateAuthenticator(i *Interaction, intent *IntentUpdateAuthenticator, step *StepState, s *State, action Action) error {
+	if step.Step != StepSetupPrimaryAuthenticator {
+		panic("interaction_update_authenticator: expected step " + step.Step)
+	}
+
+	act, ok := action.(*ActionSetupAuthenticator)
+	if !ok {
+		panic("interaction_update_authenticator: expected action type")
+	}
+
+	// Update password authenticator is the only use case for now
+	as := act.Authenticator
+	ais, err := p.Authenticator.List(i.UserID, as.Type)
+	if err != nil {
+		return err
+	}
+
+	if len(ais) == 0 {
+		return ErrAuthenticatorNotFound
+	}
+
+	ai := ais[0]
+	changed, newAuthen, err := p.Authenticator.WithSecret(i.UserID, ai, act.Secret)
```

Comment:
This first may be not a password authenticator. It is better to find the password authenticator in `ais` instead of assuming the first one is a password authenticator.

Reply:
The authenticators are fetched by type, so it must be password authenticator